src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc

```diff
@@ -42,298 +45,348 @@ namespace meshgenerator {
-
-    // Tile 3
-    for ( idx_t ix = 1; ix < N; ix++ ) ghostToOwnedTij.push_back(TijPair{{2, ix, N}, {4, 0, N - ix}});
-    for ( idx_t iy = 0; iy < N; iy++ ) ghostToOwnedTij.push_back(TijPair{{2, N, iy}, {3, 0, iy}});
-
-    // Tile 4
-    for ( idx_t ix = 0; ix < N; ix++ ) ghostToOwnedTij.push_back(TijPair{{3, ix, N}, {4, ix, 0}});
-    for ( idx_t iy = 1; iy < N; iy++ ) ghostToOwnedTij.push_back(TijPair{{3, N, iy}, {5, N - iy, 0}});
-
-    // Tile 5
-    for ( idx_t ix = 1; ix < N; ix++ ) ghostToOwnedTij.push_back(TijPair{{4, ix, N}, {0, 0, N - ix}});
-    for ( idx_t iy = 0; iy < N; iy++ ) ghostToOwnedTij.push_back(TijPair{{4, N, iy}, {5, 0, iy}});
-
-    // Tile 6
-    for ( idx_t ix = 0; ix < N; ix++ ) ghostToOwnedTij.push_back(TijPair{{5, ix, N}, {0, ix, 0}});
-    for ( idx_t iy = 1; iy < N; iy++ ) ghostToOwnedTij.push_back(TijPair{{5, N, iy}, {1, N - iy, 0}});
-
-
+void CubedSphereMeshGenerator::generate(const Grid& grid, Mesh& mesh) const {
 
-    // -------------------------------------------------------------------------
-    // END FV3 SPECIFIC MAP
-    // -------------------------------------------------------------------------
+  // Check for proper grid and need for mesh
+  ATLAS_ASSERT(!mesh.generated());
+  if (!CubedSphereGrid(grid)) {
+    throw_Exception("CubedSphereMeshGenerator can only work "
+      "with a cubedsphere grid.", Here());
+  }
 
-    ATLAS_TRACE( "CubedSphereMeshGenerator::generate" );
-    Log::debug() << "Number of faces per tile edge = " << std::to_string( N ) << std::endl;
+  // Check for proper stagger
+  const auto gridName = grid.name();
+  const auto gridStagger = gridName.substr(gridName.rfind("-") - 1, 1);
 
-    // Number of nodes
-    int nnodes     = nTiles * N * N + 2;              // Number of unique grid nodes
-    int nnodes_all = nTiles * ( N + 1 ) * ( N + 1 );  // Number of grid nodes including edge and corner duplicates
-    int ncells     = nTiles * N * N;                  // Number of unique grid cells
 
-    // Construct mesh nodes
-    // --------------------
-    mesh.nodes().resize( nnodes_all );
-    mesh::Nodes& nodes = mesh.nodes();
-    auto xy            = array::make_view<double, 2>( nodes.xy() );
-    auto lonlat        = array::make_view<double, 2>( nodes.lonlat() );
-    auto glb_idx       = array::make_view<gidx_t, 1>( nodes.global_index() );
-    auto remote_idx    = array::make_indexview<idx_t, 1>( nodes.remote_index() );
-    auto part          = array::make_view<int, 1>( nodes.partition() );
-    auto ghost         = array::make_view<int, 1>( nodes.ghost() );
-    auto flags         = array::make_view<int, 1>( nodes.flags() );
+  if (gridStagger != "C") {
+    throw_Exception("CubedSphereMeshGenerator can only work with a"
+      "cell-centroid grid. Try FV3CubedSphereMeshGenerator instead.");
+  }
 
-    // Loop over entire grid
-    // ---------------------
+  // Partitioner
+  const auto partitioner = grid::Partitioner("checkerboard", 1);
```

Comment:
I presume the above line is going to change.

Reply:
Yep, very much an hangover from the last version of this!